README
Twig lets you track progress on your topic branches.


Installation
============

    cd /usr/local/src
    git clone git://github.com/rondevera/twig
    ./install


Usage
=====

* List branches: `twig`
* Get current branch key: `twig <key>`
* Set current branch key: `twig <key> <value>`
* Get any branch key: `twig <key> -b <branch>`
* Set any branch key: `twig <key> <value> -b <branch>`


Examples
--------

List your branches, and highlight the current branch:

    $ twig

    2011-11-26 19:45:42 -0800 (6d ago)    master
    2011-11-24 17:12:09 -0800 (4d ago)    development
    2011-11-23 18:00:21 -0800 (7m ago)  * refactor_all_the_things

Set info about the current branch, e.g., which ticket it refers to. Just run
`twig <your key> <your value>`:

    $ twig issue 159

                                          issue    branch
                                          -----    ------
    2011-11-26 19:45:42 -0800 (6d ago)    -        master
    2011-11-24 17:12:09 -0800 (4d ago)    -        development
    2011-11-23 18:00:21 -0800 (7m ago)    159    * refactor_all_the_things

Show a single property of the current branch (`twig <your key>`):

    $ twig issue

    159

Set more info about the current branch (`twig <another key> <another value>`):

    $ twig status "Shipped"

                                          issue    status     branch
                                          -----    ------     ------
    2011-11-26 19:45:42 -0800 (6d ago)    -        -          master
    2011-11-24 17:12:09 -0800 (4d ago)    -        -          development
    2011-11-23 18:35:21 -0800 (3d ago)    159      Shipped  * refactor_all_the_things

Over time, track progress on multiple topic branches in parallel:

    $ twig

                                          issue    status         branch
                                          -----    ------         ------
    2011-11-26 19:45:42 -0800 (6d ago)    -        -              master
    2011-11-24 17:12:09 -0800 (4d ago)    -        -              development
    2011-11-23 18:35:21 -0800 (3d ago)    159      Shipped      * refactor_all_the_things
    2011-12-01 16:49:21 -0800 (2h ago)    268      In progress    whitespace_all_the_things
    2011-12-01 18:00:21 -0800 (7m ago)    486      In progress    optimize_all_the_things


HISTORY.md
Twig
====

* ENHANCEMENT: Support color-customizing specific property values by setting
  `<property>-style` in `.twigconfig`. For example:

      status-style: "in progress": "yellow bold", "closed": "red"

  Property values can also be styled with `--<property>-style` via command-line.
* ENHANCEMENT: Add zsh support for `twig init`. (GH-48. Thanks
  [runcom](https://github.com/runcom)!)
* ENHANCEMENT: Show warning if `~/.twigconfig` contains an invalid option.
  (GH-40)
* FIX: Unescape branch properties on display. (GH-52. Thanks [rcoder](https://github.com/rcoder)!)

1.7.1 (2015-08-19)
------------------
* ENHANCEMENT: Add Ruby 2.2 to official list of supported Rubies. (GH-47. Thanks
  [apuratepp](https://github.com/apuratepp)!)

1.7 (2014-07-29)
----------------
* ENHANCEMENT: Expand `twig init` to also create a default `~/.twigconfig` file.
  This uses a new `twig init-config` subcommand behind the scenes.
* ENHANCEMENT: Add `twig help <subcommand>` for showing help content for any
  built-in Twig subcommand (e.g., `twig help rebase`). To make a custom
  subcommand work with this system (e.g., `twig help my-subcommand`), simply add
  `--help` support to the subcommand (e.g., `twig my-subcommand --help`), and
  Twig will call it automatically.
* ENHANCEMENT: List built-in subcommands in `twig help` with short descriptions
  of each. (GH-32)
* ENHANCEMENT: Support `twig rebase --autoconfirm` for skipping the prompt when
  rebasing a branch onto its parent branch. To skip the prompt every time you
  use `twig rebase`, add `twig-rebase-autoconfirm: true` to your
  `~/.twigconfig`. (GH-24. Thanks [ljharb](https://github.com/ljharb) for the
  idea!)
* ENHANCEMENT: Suggest `twig --all` if there are no matching branches. (GH-33)
* ENHANCEMENT: Add Ruby 2.1.1 and 2.1.2 to official list of supported Rubies.
* FIX: Don't require a Git repo for initialization (e.g., `twig init`) or for
  showing help content (e.g., `twig help rebase`).
* FIX: Show units of time for relative time strings when not using English.
  (GH-37, GH-38. Thanks [matteosister](https://github.com/matteosister)!)
* FIX: Handle special shell characters (e.g., backticks, dollar signs) in branch
  names and property values.
* FIX: Disable unsupported string formatting in Windows. (GH-39. Thanks
  [rcdailey](https://github.com/rcdailey) for the bug report!)

1.6 (2014-03-04)
----------------
* ENHANCEMENT: Add `--only-property <pattern>` and `--except-property <pattern>`
  for including/excluding property names in list view and JSON view. Useful for
  hiding properties that are frequently too long to show in list view (e.g.,
  `todo`), or for hiding groups of properties that are just issue tracker
  metadata. (GH-25. Thanks [slucero](https://github.com/slucero) for the idea!)
* ENHANCEMENT: Add `twig init` subcommand, the new recommended way to set up
  Twig after installing the gem.
* ENHANCEMENT: Add `twig checkout-parent` subcommand, which checks out the
  current branch's parent branch, if any, based on its `diff-branch` property.
  (GH-34)
* ENHANCEMENT: Add `twig checkout-child` subcommand, which checks out the
  current branch's child branch, if any, based on the child's `diff-branch`
  property. If the current branch has multiple child branches, this subcommand
  lists all of them and prompts for a selection. (GH-35)
* ENHANCEMENT: Add `twig create-branch` subcommand, which creates a branch off
  of the current branch, and sets the child branch's `diff-branch` property
  automatically.
* FIX: Fix `(1y ago)` (previously displayed as `(1 ago)`) in branch list view.
* FIX: Make `property` a reserved property name, along with `branch`, `merge`,
  `rebase`, and `remote`.

1.5 (2013-11-21)
----------------
* ENHANCEMENT: Add `--format=json` option for printing branch data as JSON
  instead of a list. Useful for integrating Twig data into other tools.
* ENHANCEMENT: Add tab completion for all subcommands, built-in (e.g., `twig
  diff`, `twig gh-update`) and custom.
* ENHANCEMENT: Paginate help content where possible.
* ENHANCEMENT: Improve error messages `~/.twigconfig` isn't readable or contains
  invalid lines.
* ENHANCEMENT: Include default option values for branch listing and GitHub
  integration in `twig --help`. (GH-30, GH-31)
* FIX: Fix warnings when listing branches when a branch has UTF-8 characters in
  its name. (GH-20)
* FIX: Fix showing relative time for very old branches (e.g., "2y ago"). (GH-29)

1.4 (2013-08-07)
----------------
* ENHANCEMENT: Speed up listing branches by 3–4x.
* ENHANCEMENT: Add `twig diff`, which diffs the current branch against its
  `diff-branch` property, and `twig diff <branch>`, which diffs the given branch
  against its `diff-branch` property.
* ENHANCEMENT: Add `twig rebase`, which rebases the current branch onto the
  branch in its `diff-branch` property, and `twig rebase <branch>`, which
  rebases the given branch onto its `diff-branch`.
* ENHANCEMENT: Add `twig init-completion --force` for overwriting existing
  completion scripts. Useful for upgrading to the latest Twig completion
  scripts.
* ENHANCEMENT: Improve `twig gh-update` error reporting by listing the affected
  API endpoints upon failure.
* FIX: Deprecate `~/.twigrc` in favor of `~/.twigconfig`.

1.3 (2013-05-22)
----------------
* ENHANCEMENT: Add `--branch-width` and `--<property>-width` options for setting
  custom column widths.
* ENHANCEMENT: Add `--reverse` option for listing least recently updated
  branches first. This can be used in a config file as `reverse: true`.
* ENHANCEMENT: Make `gh-open` and `gh-open-issue` work cross-platform.
  (GH-18. Thanks [ixti](https://github.com/ixti)!)
* FIX: Allow getting, setting, and unsetting properties for branches older than
  the `max-days-old` option, if given.
* FIX: Abort `twig gh-*` subcommands early if working in a non-Github
  repository.

1.2.1 (2013-05-04)
------------------
* FIX: Add User-Agent string to `twig gh-update` GitHub requests to comply with

lib/twig/cli/help.rb
class Twig
  module Cli
    # Handles printing help output for `twig help`.
    module Help
      def self.console_width
        80
      end

      def self.intro
        version_string = "Twig v#{Twig::VERSION}"

        intro = Help.paragraph(%{
          Twig is your personal Git branch assistant. It's a command-line tool
          for listing your most recent branches, and for remembering branch
          details for you, like issue tracker ids and todos. It also supports
          subcommands, like automatically fetching statuses from your issue
          tracking system.
        })

        intro = <<-BANNER.gsub(/^[ ]+/, '')

          #{'=' * version_string.size}
          #{version_string}
          #{'=' * version_string.size}

          #{intro}

          #{Twig::HOMEPAGE}
        BANNER

        intro + ' ' # Force extra blank line
      end

      def self.description(text, options = {})
        defaults = {
          :add_blank_line => false,
          :width => 40
        }
        options = defaults.merge(options)

        width = options[:width]
        words = text.gsub(/\n?\s+/, ' ').strip.split(' ')
        lines = []

        # Split words into lines
        while words.any?
          current_word      = words.shift
          current_word_size = Display.unformat_string(current_word).size
          last_line         = lines.last
          last_line_size    = last_line && Display.unformat_string(last_line).size

          if last_line_size && (last_line_size + current_word_size + 1 <= width)
            last_line << ' ' << current_word
          elsif current_word_size >= width
            lines << current_word[0...width]
            words.unshift(current_word[width..-1])
          else
            lines << current_word
          end
        end

        lines << ' ' if options[:add_blank_line]
        lines
      end

      def self.description_for_custom_property(option_parser, desc_lines, options = {})
        options[:trailing] ||= "\n"
        indent = '      '
        left_column_width = 29

        help_desc = desc_lines.inject('') do |desc, (left_column, right_column)|
          desc + indent +
          sprintf("%-#{left_column_width}s", left_column) + right_column + "\n"
        end

        Help.print_section(option_parser, help_desc, :trailing => options[:trailing])
      end

      def self.line_for_custom_property?(line)
        is_custom_property_except = (
          line.include?('--except-') &&
          !line.include?('--except-branch') &&
          !line.include?('--except-property') &&
          !line.include?('--except-PROPERTY')
        )
        is_custom_property_only = (
          line.include?('--only-') &&
          !line.include?('--only-branch') &&
          !line.include?('--only-property') &&
          !line.include?('--only-PROPERTY')
        )
        is_custom_property_width = (
          line =~ /--.+-width/ &&
          !line.include?('--branch-width') &&
          !line.include?('--PROPERTY-width')
        )

        is_custom_property_except ||
        is_custom_property_only ||
        is_custom_property_width
      end

      def self.paragraph(text)
        Help.description(text, :width => console_width).join("\n")
      end

      def self.print_line(option_parser, text)
        # Prints a single line of text without line breaks.
        option_parser.separator(text)
      end

      def self.print_paragraph(option_parser, text, separator_options = {})
        # Prints a long chunk of text with automatic word wrapping and a leading
        # line break.

        separator_options[:trailing] ||= ''
        Help.print_section(option_parser, Help.paragraph(text), separator_options)
      end

      def self.print_section(option_parser, text, options = {})
        # Prints text with leading and trailing line breaks.

        options[:trailing] ||= ''
        option_parser.separator "\n#{text}#{options[:trailing]}"
      end

      def self.subcommand_descriptions
        descs = {
          'checkout-child'  => 'Checks out a branch\'s child branch, if any.',
          'checkout-parent' => 'Checks out a branch\'s parent branch.',
          'create-branch'   => 'Creates a branch and sets its `diff-branch` property to the previous branch name.',
          'diff'            => 'Shows the diff between a branch and its parent branch (`diff-branch`).',
          'gh-open'         => 'Opens a browser window for the current GitHub repository.',
          'gh-open-issue'   => 'Opens a browser window for a branch\'s GitHub issue, if any.',
          'gh-update'       => 'Updates each branch with the latest issue status on GitHub.',
          'help'            => 'Provides help for Twig and its subcommands.',
          'init'            => 'Runs all Twig setup commands.',
          'init-completion' => 'Initializes tab completion for Twig. Runs as part of `twig init`.',
          'init-config'     => 'Creates a default `~/.twigconfig` file. Runs as part of `twig init`.',
          'rebase'          => 'Rebases a branch onto its parent branch (`diff-branch`).'
        }

        line_prefix    = '- '
        gutter_width   = 2 # Space between columns
        names          = descs.keys.sort
        max_name_width = names.map { |name| name.length }.max
        names_width    = max_name_width + gutter_width
        descs_width    = Help.console_width - line_prefix.length - names_width
        desc_indent    = ' ' * (names_width + line_prefix.length)

        names.map do |name|
          line_prefix +
          sprintf("%-#{names_width}s", name) +
          Help.description(descs[name], :width => descs_width).join("\n" + desc_indent)
        end
      end

      def self.header(option_parser, text, separator_options = {}, header_options = {})
        separator_options[:trailing] ||= "\n\n"
        header_options[:underline]   ||= '='

        Help.print_section(
          option_parser,
          text + "\n" + (header_options[:underline] * text.size),
          separator_options
        )
      end

      def self.subheader(option_parser, text, separator_options = {})
        header(option_parser, text, separator_options, :underline => '-')
      end
    end
  end
end
